larpandoracontent/LArObjects/LArAdaBoostDecisionTree.cc

```diff
@@ -0,0 +1,394 @@
+
+    WeakClassifiers weakClassifiers;
+    const pandora::StatusCode statusCode(this->ReadXmlFile(parameterLocation, bdtName, weakClassifiers));
+
+    if (pandora::STATUS_CODE_SUCCESS != statusCode)
+    {
+        this->Reset();
+        std::cout << "AdaBoostDecisionTree: Unable to read xml" << std::endl;
+        throw pandora::StatusCodeException(statusCode);
+    }
+
+    // ATTN: Add consistency check for number of features, bearing in min not all features in a bdt may be used
+    m_pStrongClassifier = new StrongClassifier(weakClassifiers);
+
+    return pandora::STATUS_CODE_SUCCESS;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void AdaBoostDecisionTree::Reset() const
+{
+    if (!m_pStrongClassifier)
+        return;
+
+    for (const WeakClassifier *pWeakClassifier : m_pStrongClassifier->GetWeakClassifiers())
+    {
+        if (!pWeakClassifier)
+            continue;
+
+        pWeakClassifier->Reset();
+        delete pWeakClassifier;
+    }
+
+    delete m_pStrongClassifier;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+bool AdaBoostDecisionTree::Classify(const DoubleVector &features) const
+{
+    return (this->CalculateScore(features) > 0.0 ? true : false);
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+double AdaBoostDecisionTree::CalculateClassificationScore(const DoubleVector &features) const
+{
+    return this->CalculateScore(features);
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+double AdaBoostDecisionTree::CalculateProbability(const DoubleVector &features) const
+{
+    // ATTN: BDT score, once normalised by total weight, is confined to the range -1 to +1.  This linear mapping places the score in the 
+    // range 0 to 1 so that it may be interpreted as a probability. 
+    return (this->CalculateScore(features) + 1.0) * 0.5;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+double AdaBoostDecisionTree::CalculateScore(const DoubleVector &features) const
+{
+    if (m_pStrongClassifier == nullptr)
+    {
+        this->Reset();
+        std::cout << "AdaBoostDecisionTree: Attempting to use an uninitialized bdt" << std::endl;
+        throw pandora::StatusCodeException(pandora::STATUS_CODE_NOT_INITIALIZED);
+    }
+
+    double score(0.0);
+
+    try
+    {
+        score = m_pStrongClassifier->Predict(features);
+    }
+    catch (pandora::StatusCodeException &statusCodeException)
+    {
+        if (pandora::STATUS_CODE_NOT_FOUND == statusCodeException.GetStatusCode())
+        {
+            this->Reset();
```

Comment:
Are you sure that you want to rip out everything in the classifier if you get an exception on one event? More typically, just need to ensure that everything is ready/reset to attempt to process next event without things being left in a mess.